ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/AutoMediaBrowserService.kt

```diff
@@ -0,0 +1,186 @@
+package org.moire.ultrasonic.service
+
+import android.os.Bundle
+import android.os.Handler
+import android.support.v4.media.MediaBrowserCompat
+import android.support.v4.media.MediaDescriptionCompat
+import android.support.v4.media.session.MediaSessionCompat
+import androidx.media.MediaBrowserServiceCompat
+import androidx.media.utils.MediaConstants
+import org.koin.android.ext.android.inject
+import org.moire.ultrasonic.util.MediaSessionEventDistributor
+import org.moire.ultrasonic.util.MediaSessionEventListener
+import timber.log.Timber
+
+
+const val MY_MEDIA_ROOT_ID = "MY_MEDIA_ROOT_ID"
+const val MY_MEDIA_ALBUM_ID = "MY_MEDIA_ALBUM_ID"
+const val MY_MEDIA_ARTIST_ID = "MY_MEDIA_ARTIST_ID"
+const val MY_MEDIA_ALBUM_ITEM = "MY_MEDIA_ALBUM_ITEM"
+const val MY_MEDIA_LIBRARY_ID = "MY_MEDIA_LIBRARY_ID"
+const val MY_MEDIA_PLAYLIST_ID = "MY_MEDIA_PLAYLIST_ID"
+
+class AutoMediaBrowserService : MediaBrowserServiceCompat() {
+
+    private lateinit var mediaSessionEventListener: MediaSessionEventListener
+    private val mediaSessionEventDistributor: MediaSessionEventDistributor by inject()
+    private val lifecycleSupport: MediaPlayerLifecycleSupport by inject()
+
+    override fun onCreate() {
+        super.onCreate()
+
+        mediaSessionEventListener = object : MediaSessionEventListener {
+            override fun onMediaSessionTokenCreated(token: MediaSessionCompat.Token) {
+                Timber.i("AutoMediaBrowserService onMediaSessionTokenCreated called")
+                if (sessionToken == null) {
+                    Timber.i("AutoMediaBrowserService onMediaSessionTokenCreated session token was null, set it to %s", token.toString())
+                    sessionToken = token
+                }
+            }
+
+            override fun onPlayFromMediaIdRequested(mediaId: String?, extras: Bundle?) {
+                // TODO implement
+                Timber.i("AutoMediaBrowserService onPlayFromMediaIdRequested called")
+            }
+
+            override fun onPlayFromSearchRequested(query: String?, extras: Bundle?) {
+                // TODO implement
+                Timber.i("AutoMediaBrowserService onPlayFromSearchRequested called")
+            }
+        }
+
+        mediaSessionEventDistributor.subscribe(mediaSessionEventListener)
+
+        val handler = Handler()
+        handler.postDelayed({
+            Timber.i("AutoMediaBrowserService starting lifecycleSupport and MediaPlayerService...")
+            // TODO it seems Android Auto handles autostart, but we must check that
```

Comment:
So this here has a big problem.
I'm not sure how can we make the AndroidAuto logic to work with how Ultrasonic works currently.

When Ultrasonic is started, it usually loads up a playlist, and resumes playback if it wasn't completely stopped before.
If it was stopped, the playback can be started from the Now Playing screen.

However, when Ultrasonic is stopped, the Now Playing is not visible at all in Android Auto! I think because it uses the MediaSession, which is without a current track, and inactive in this case. So in Android Auto the playback only can be started after selecting a new something to play.

To solve this nicely, I think we must rethink how this should work and maybe refactor the MediaSession handling.

Reply:
MediaSession handling is now refactored. Starting the MediaPlayerService and LifecycleSupport is still necessary, when Ultrasonic is booted up by Android Auto